Don't log byte arrays in CommandDecoder

Hi,
we sometimes run into OutOfMemoryException while Redisson is trying to log a warning, since it includes the content in the log message [1]. Would it be possible to check for the instance type and in case of byte[] just log something else? I.e.
`
log.warn("response has been skipped due to timeout! channel: {}, command: {}, result: {}",
                         channel, data, result instanceof byte[] ? result.hashCode() : result);
`
Or omit the content in the WARN-log and do a second INFO-log containing the data. So you can toggle this via the log level?
`
log.warn("response has been skipped due to timeout! channel: {}, command: {}", channel, data);
log.info("skipped response data for timeout {}", result); 
`
From what I can see that's the only place where large content might be locked, though I am not sure about the log.error in CommandDecoder#decode, as it is using the CommandData.toString() content. 

Thanks,
Philipp

[1]
`
An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.handler.codec.DecoderException: java.lang.OutOfMemoryError: Java heap space
    at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:422)
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351)
    at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:351)
    at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:373)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
    at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:1018)
    at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:402)
    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:307)
    at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.OutOfMemoryError: Java heap space
    at java.util.Arrays.copyOf(Arrays.java:3332)
    at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
    at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
    at java.lang.StringBuilder.append(StringBuilder.java:136)
    at org.slf4j.helpers.MessageFormatter.byteArrayAppend(MessageFormatter.java:342)
    at org.slf4j.helpers.MessageFormatter.deeplyAppendParameter(MessageFormatter.java:276)
    at org.slf4j.helpers.MessageFormatter.arrayFormat(MessageFormatter.java:227)
    at org.ops4j.pax.logging.slf4j.Slf4jLogger.warn(Slf4jLogger.java:715)
    at org.redisson.client.handler.CommandDecoder.handleResult(CommandDecoder.java:384)
    at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:283)
    at org.redisson.client.handler.CommandDecoder.decode(CommandDecoder.java:125)
    at io.netty.handler.codec.ReplayingDecoder.callDecode(ReplayingDecoder.java:367)
    ... 18 more
`

--- a/redisson/src/main/java/org/redisson/client/handler/CommandDecoder.java
+++ b/redisson/src/main/java/org/redisson/client/handler/CommandDecoder.java
@@ -45,6 +45,7 @@ import org.redisson.client.protocol.pubsub.Message;
 import org.redisson.client.protocol.pubsub.PubSubMessage;
 import org.redisson.client.protocol.pubsub.PubSubPatternMessage;
 import org.redisson.client.protocol.pubsub.PubSubStatusMessage;
+import org.redisson.misc.LogHelper;
 import org.redisson.misc.RPromise;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -203,11 +204,11 @@ public class CommandDecoder extends ReplayingDecoder<State> {
             RPromise<Void> promise = commandBatch.getPromise();
             if (error != null) {
                 if (!promise.tryFailure(error) && promise.cause() instanceof RedisTimeoutException) {
-                    log.warn("response has been skipped due to timeout! channel: {}, command: {}", ctx.channel(), data);
+                    log.warn("response has been skipped due to timeout! channel: {}, command: {}",ctx.channel(), LogHelper.toString(data));
                 }
             } else {
                 if (!promise.trySuccess(null) && promise.cause() instanceof RedisTimeoutException) {
-                    log.warn("response has been skipped due to timeout! channel: {}, command: {}", ctx.channel(), data);
+                    log.warn("response has been skipped due to timeout! channel: {}, command: {}", ctx.channel(), LogHelper.toString(data));
                 }
             }
             

--- a/redisson/src/main/java/org/redisson/client/protocol/CommandData.java
+++ b/redisson/src/main/java/org/redisson/client/protocol/CommandData.java
@@ -21,6 +21,7 @@ import java.util.List;
 
 import org.redisson.client.codec.Codec;
 import org.redisson.client.protocol.decoder.MultiDecoder;
+import org.redisson.misc.LogHelper;
 import org.redisson.misc.RPromise;
 
 /**
@@ -85,7 +86,7 @@ public class CommandData<T, R> implements QueueCommand {
     @Override
     public String toString() {
         return "CommandData [promise=" + promise + ", command=" + command + ", params="
-                + Arrays.toString(params) + ", codec=" + codec + "]";
+                + LogHelper.toString(params) + ", codec=" + codec + "]";
     }
 
     @Override

--- a/redisson/src/main/java/org/redisson/command/CommandAsyncService.java
+++ b/redisson/src/main/java/org/redisson/command/CommandAsyncService.java
@@ -49,6 +49,7 @@ import org.redisson.connection.ConnectionManager;
 import org.redisson.connection.MasterSlaveEntry;
 import org.redisson.connection.NodeSource;
 import org.redisson.connection.NodeSource.Redirect;
+import org.redisson.misc.LogHelper;
 import org.redisson.misc.RPromise;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -520,7 +521,7 @@ public class CommandAsyncService implements CommandAsyncExecutor {
 
                 if (details.getAttempt() == connectionManager.getConfig().getRetryAttempts()) {
                     if (details.getException() == null) {
-                        details.setException(new RedisTimeoutException("Command execution timeout for command: " + command + " with params: " + Arrays.toString(details.getParams())));
+                        details.setException(new RedisTimeoutException("Command execution timeout for command: " + command + " with params: " + LogHelper.toString(details.getParams())));
                     }
                     details.getAttemptPromise().tryFailure(details.getException());
                     return;
@@ -605,7 +606,7 @@ public class CommandAsyncService implements CommandAsyncExecutor {
 
         if (!future.isSuccess()) {
             details.setException(new WriteRedisConnectionException(
-                    "Can't write command: " + details.getCommand() + ", params: " + Arrays.toString(details.getParams()) + " to channel: " + future.channel(), future.cause()));
+                    "Can't write command: " + details.getCommand() + ", params: " + LogHelper.toString(details.getParams()) + " to channel: " + future.channel(), future.cause()));
             return;
         }
 
@@ -629,7 +630,7 @@ public class CommandAsyncService implements CommandAsyncExecutor {
             public void run(Timeout timeout) throws Exception {
                 details.getAttemptPromise().tryFailure(
                         new RedisTimeoutException("Redis server response timeout (" + timeoutAmount + " ms) occured for command: " + details.getCommand()
-                                + " with params: " + Arrays.toString(details.getParams()) + " channel: " + connection.getChannel()));
+                                + " with params: " + LogHelper.toString(details.getParams()) + " channel: " + connection.getChannel()));
             }
         };
 

--- a/redisson/src/main/java/org/redisson/misc/LogHelper.java
+++ b/redisson/src/main/java/org/redisson/misc/LogHelper.java
@@ -1,0 +1,81 @@
+/**
+ * Copyright 2016 Nikita Koksharov
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.redisson.misc;
+
+import java.lang.reflect.Array;
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * @author Philipp Marx
+ */
+public class LogHelper {
+
+    private static final int MAX_COLLECTION_LOG_SIZE = Integer.valueOf(System.getProperty("redisson.maxCollectionLogSize", "10"));
+    private static final int MAX_STRING_LOG_SIZE = Integer.valueOf(System.getProperty("redisson.maxStringLogSize", "100"));
+
+    private LogHelper() {
+    }
+
+    public static <T> String toString(T object) {
+        if (object == null) {
+            return "null";
+        } else if (object instanceof String) {
+            return toStringString((String) object);
+        } else if (object.getClass().isArray()) {
+            return toArrayString(object);
+        } else if (object instanceof Collection) {
+            return toCollectionString((Collection<?>) object);
+        } else {
+            return Objects.toString(object);
+        }
+    }
+
+    private static String toStringString(String string) {
+        if (string.length() > MAX_STRING_LOG_SIZE) {
+            return new StringBuilder(string.substring(0, MAX_STRING_LOG_SIZE)).append("...").toString();
+        } else {
+            return string;
+        }
+    }
+
+    private static String toCollectionString(Collection<?> collection) {
+        return toArrayString(collection.toArray(new Object[collection.size()]));
+    }
+
+    private static String toArrayString(Object array) {
+        int length = Array.getLength(array) - 1;
+        if (length == -1) {
+            return "[]";
+        }
+
+        StringBuilder b = new StringBuilder(length * 3);
+        b.append('[');
+        for (int i = 0;; ++i) {
+            b.append(toString(Array.get(array, i)));
+
+            if (i == length) {
+                return b.append(']').toString();
+            }
+
+            b.append(", ");
+
+            if (i == MAX_COLLECTION_LOG_SIZE - 1) {
+                return b.append("...]").toString();
+            }
+        }
+    }
+}

--- a/redisson/src/test/java/org/redisson/misc/LogHelperTest.java
+++ b/redisson/src/test/java/org/redisson/misc/LogHelperTest.java
@@ -1,0 +1,223 @@
+package org.redisson.misc;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.Test;
+
+/**
+ * @author Philipp Marx
+ */
+public class LogHelperTest {
+
+    @Test
+    public void toStringWithNull() {
+        assertThat(LogHelper.toString(null), is("null"));
+    }
+
+    @Test
+    public void toStringWithNestedPrimitives() {
+        Object[] input = new Object[] { "0", 1, 2L, 3.1D, 4.2F, (byte) 5, '6' };
+
+        assertThat(LogHelper.toString(input), is("[0, 1, 2, 3.1, 4.2, 5, 6]"));
+    }
+
+    @Test
+    public void toStringWithPrimitive() {
+        assertThat(LogHelper.toString("0"), is("0"));
+        assertThat(LogHelper.toString(1), is("1"));
+        assertThat(LogHelper.toString(2L), is("2"));
+        assertThat(LogHelper.toString(3.1D), is("3.1"));
+        assertThat(LogHelper.toString(4.2F), is("4.2"));
+        assertThat(LogHelper.toString((byte) 5), is("5"));
+        assertThat(LogHelper.toString('6'), is("6"));
+    }
+
+    @Test
+    public void toStringWithNestedSmallArrays() {
+        String[] strings = new String[] { "0" };
+        int[] ints = new int[] { 1 };
+        long[] longs = new long[] { 2L };
+        double[] doubles = new double[] { 3.1D };
+        float[] floats = new float[] { 4.2F };
+        byte[] bytes = new byte[] { (byte) 5 };
+        char[] chars = new char[] { '6' };
+
+        Object[] input = new Object[] { strings, ints, longs, doubles, floats, bytes, chars };
+
+        assertThat(LogHelper.toString(input), is("[[0], [1], [2], [3.1], [4.2], [5], [6]]"));
+    }
+    
+    @Test
+    public void toStringWithNestedSmallCollections() {
+        List<String> strings = Arrays.asList("0" );
+        List<Integer> ints = Arrays.asList( 1 );
+        List<Long> longs = Arrays.asList( 2L );
+        List<Double> doubles = Arrays.asList( 3.1D );
+        List<Float> floats = Arrays.asList( 4.2F );
+        List<Byte> bytes = Arrays.asList( (byte) 5 );
+        List<Character> chars = Arrays.asList( '6' );
+        
+        Object[] input = new Object[] { strings, ints, longs, doubles, floats, bytes, chars };
+        
+        assertThat(LogHelper.toString(input), is("[[0], [1], [2], [3.1], [4.2], [5], [6]]"));
+    }
+    
+    @Test
+    public void toStringWithSmallArrays() {
+        String[] strings = new String[] { "0" };
+        int[] ints = new int[] { 1 };
+        long[] longs = new long[] { 2L };
+        double[] doubles = new double[] { 3.1D };
+        float[] floats = new float[] { 4.2F };
+        byte[] bytes = new byte[] { (byte) 5 };
+        char[] chars = new char[] { '6' };
+
+        assertThat(LogHelper.toString(strings), is("[0]"));
+        assertThat(LogHelper.toString(ints), is("[1]"));
+        assertThat(LogHelper.toString(longs), is("[2]"));
+        assertThat(LogHelper.toString(doubles), is("[3.1]"));
+        assertThat(LogHelper.toString(floats), is("[4.2]"));
+        assertThat(LogHelper.toString(bytes), is("[5]"));
+        assertThat(LogHelper.toString(chars), is("[6]"));
+    }
+    
+    @Test
+    public void toStringWithSmallCollections() {
+        List<String> strings = Collections.nCopies(1, "0");
+        List<Integer> ints =  Collections.nCopies(1, 1);
+        List<Long> longs =  Collections.nCopies(1, 2L);
+        List<Double> doubles =  Collections.nCopies(1, 3.1D);
+        List<Float> floats =  Collections.nCopies(1, 4.2F);
+        List<Byte> bytes =  Collections.nCopies(1, (byte)5);
+        List<Character> chars =  Collections.nCopies(1, '6');
+        
+        assertThat(LogHelper.toString(strings), is("[0]"));
+        assertThat(LogHelper.toString(ints), is("[1]"));
+        assertThat(LogHelper.toString(longs), is("[2]"));
+        assertThat(LogHelper.toString(doubles), is("[3.1]"));
+        assertThat(LogHelper.toString(floats), is("[4.2]"));
+        assertThat(LogHelper.toString(bytes), is("[5]"));
+        assertThat(LogHelper.toString(chars), is("[6]"));
+    }
+
+    @Test
+    public void toStringWithNestedBigArrays() {
+        String[] strings = new String[15];
+        Arrays.fill(strings, "0");
+        int[] ints = new int[15];
+        Arrays.fill(ints, 1);
+        long[] longs = new long[15];
+        Arrays.fill(longs, 2L);
+        double[] doubles = new double[15];
+        Arrays.fill(doubles, 3.1D);
+        float[] floats = new float[15];
+        Arrays.fill(floats, 4.2F);
+        byte[] bytes = new byte[15];
+        Arrays.fill(bytes, (byte) 5);
+        char[] chars = new char[15];
+        Arrays.fill(chars, '6');
+
+        Object[] input = new Object[] { strings, ints, longs, doubles, floats, bytes, chars };
+        StringBuilder sb = new StringBuilder();
+        sb.append("[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...], ");
+        sb.append("[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...], ");
+        sb.append("[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, ...], ");
+        sb.append("[3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, ...], ");
+        sb.append("[4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, ...], ");
+        sb.append("[5, 5, 5, 5, 5, 5, 5, 5, 5, 5, ...], ");
+        sb.append("[6, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...]]");
+
+        assertThat(LogHelper.toString(input), is(sb.toString()));
+    }
+    
+    @Test
+    public void toStringWithNestedBigCollections() {
+        List<String> strings = Collections.nCopies(15, "0");
+        List<Integer> ints =  Collections.nCopies(15, 1);
+        List<Long> longs =  Collections.nCopies(15, 2L);
+        List<Double> doubles =  Collections.nCopies(15, 3.1D);
+        List<Float> floats =  Collections.nCopies(15, 4.2F);
+        List<Byte> bytes =  Collections.nCopies(15, (byte)5);
+        List<Character> chars =  Collections.nCopies(15, '6');
+        
+        Object[] input = new Object[] { strings, ints, longs, doubles, floats, bytes, chars };
+        StringBuilder sb = new StringBuilder();
+        sb.append("[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...], ");
+        sb.append("[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...], ");
+        sb.append("[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, ...], ");
+        sb.append("[3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, ...], ");
+        sb.append("[4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, ...], ");
+        sb.append("[5, 5, 5, 5, 5, 5, 5, 5, 5, 5, ...], ");
+        sb.append("[6, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...]]");
+        
+        assertThat(LogHelper.toString(input), is(sb.toString()));
+    }
+
+    @Test
+    public void toStringWithBigArrays() {
+        String[] strings = new String[15];
+        Arrays.fill(strings, "0");
+        int[] ints = new int[15];
+        Arrays.fill(ints, 1);
+        long[] longs = new long[15];
+        Arrays.fill(longs, 2L);
+        double[] doubles = new double[15];
+        Arrays.fill(doubles, 3.1D);
+        float[] floats = new float[15];
+        Arrays.fill(floats, 4.2F);
+        byte[] bytes = new byte[15];
+        Arrays.fill(bytes, (byte) 5);
+        char[] chars = new char[15];
+        Arrays.fill(chars, '6');
+
+        assertThat(LogHelper.toString(strings), is("[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...]"));
+        assertThat(LogHelper.toString(ints), is("[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...]"));
+        assertThat(LogHelper.toString(longs), is("[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, ...]"));
+        assertThat(LogHelper.toString(doubles), is("[3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, ...]"));
+        assertThat(LogHelper.toString(floats), is("[4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, ...]"));
+        assertThat(LogHelper.toString(bytes), is("[5, 5, 5, 5, 5, 5, 5, 5, 5, 5, ...]"));
+        assertThat(LogHelper.toString(chars), is("[6, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...]"));
+    }
+    
+    @Test
+    public void toStringWithBigCollections() {
+        List<String> strings = Collections.nCopies(15, "0");
+        List<Integer> ints =  Collections.nCopies(15, 1);
+        List<Long> longs =  Collections.nCopies(15, 2L);
+        List<Double> doubles =  Collections.nCopies(15, 3.1D);
+        List<Float> floats =  Collections.nCopies(15, 4.2F);
+        List<Byte> bytes =  Collections.nCopies(15, (byte)5);
+        List<Character> chars =  Collections.nCopies(15, '6');
+        
+        assertThat(LogHelper.toString(strings), is("[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...]"));
+        assertThat(LogHelper.toString(ints), is("[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...]"));
+        assertThat(LogHelper.toString(longs), is("[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, ...]"));
+        assertThat(LogHelper.toString(doubles), is("[3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, 3.1, ...]"));
+        assertThat(LogHelper.toString(floats), is("[4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, 4.2, ...]"));
+        assertThat(LogHelper.toString(bytes), is("[5, 5, 5, 5, 5, 5, 5, 5, 5, 5, ...]"));
+        assertThat(LogHelper.toString(chars), is("[6, 6, 6, 6, 6, 6, 6, 6, 6, 6, ...]"));
+    }
+
+    @Test
+    public void toStringWithSmallString() {
+        char[] charsForStr = new char[100];
+        Arrays.fill(charsForStr, '7');
+        String string = new String(charsForStr);
+
+        assertThat(LogHelper.toString(string), is(string));
+    }
+
+    @Test
+    public void toStringWithBigString() {
+        char[] charsForStr = new char[150];
+        Arrays.fill(charsForStr, '7');
+        String string = new String(charsForStr);
+
+        assertThat(LogHelper.toString(string), is(string.substring(0, 100) + "..."));
+    }
+}

